3D CylinderComponent

lipohlmann opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
closes #31157

## Reason
Needing to test 3D extrusion capabilities for future mesh connector. Also, 3D pipes are better than 2D...

## Design
Users can create a 3D cylinder with a specified center point and extrusion direction (i.e. which way is "up"). Utilizes TransformGenerator to get the user-specified orientations correct. Expands on the CylinderComponent ActionComponent.

## Impact
Adds new capabilities in the form of an improved ActionComponent.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

Commit:
Add getters for mesh transformations of the component in Helper class

Commit:
Begin modifications to CylinderComponent to allow for 3D cylinders. Need to address issue with Seg Fault with attempting to get the direciton.

Commit:
Fix rotation angle computation in ComponentMeshTransformHelper, center Cylinder component on prescribed point

Commit:
Add mooseAsserts to check results.

Commit:
EOD 7/28. Orientation of meshes still causing issues. Need to look into conversion from XYZ extrinsic rotation to XZX intrinsic rotation (see TranformGenerator)

Commit:
First working draft of 3D CylinderComponent.

 - Added new method for creating rotation matrix between two vectors.
 - Updated ComponentMeshTransformHelper to rotate correctly.
 - Added new capabilities for rotation in TransformGenerator using either extrinsic angles or a given rotation matrix.
 - Created simple test file for verifying everything works.

closes #31157

Commit:
Write unit test for new rotation matrix method. Change name for clarity.

Commit:
Create tests and gold files.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3081744/), step Docs: sync website on cee3c82 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31158/site)

This comment will be updated on new commits.

Commit:
Fix some parameter issues. Create boundary names for 3D cylinders.

joshuahansel commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -38,9 +38,20 @@ CylinderComponent::validParams()
   params.addRequiredRangeCheckedParam<Real>("radius", "radius>0", "Radius of the cylinder");
   params.addRequiredRangeCheckedParam<Real>("length", "length>0", "Length/Height of the cylinder");
 
+  params.addParam<unsigned int>("n_radial",
+                                "Number of radial elements. Only assign if mesh is 2D!");
+  params.addParam<unsigned int>(
+      "n_radial_rings",
+      1,
+      "Number of radial rings within the mesh (does not affect boundary layers).");
```
What's the difference between a "ring" and radial element?

joshuahansel commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -38,9 +38,20 @@ CylinderComponent::validParams()
   params.addRequiredParam<unsigned int>("n_axial", "Number of axial elements of the cylinder");
-  params.addParam<unsigned int>("n_radial", "Number of radial elements of the cylinder");
-  params.addParam<unsigned int>("n_azimuthal", "Number of azimuthal elements of the cylinder");
+  params.addRangeCheckedParam<Real>("boundary_layer_width",
```
Are boundary layers applied to both axial and radial boundaries?

joshuahansel replied:
Just an idea, which you can decline: consider instead of hard-wiring this "boundary layer" just giving the user total freedom of radial regions and numbers of elements in each.

joshuahansel reviewed (changes requested):
Just passing by to leave a few comments. For the user questions I had, please make sure you modify documentation accordingly to clarify.

Commit:
Fix bug in creating the Rodigues rotation matrix where the angle was being calculated with the wrong vectors.

GiudGiud committed:
Safer component mesh transform
More explanation on the role of direction

GiudGiud committed:
Address review:
- organize parameters more
- simplify rings vs radial rings idea
- no -Y component in 2D cylinder, because solved in RZ axis
- add RZ axis setting: note we can only handle a singe cylinder now
- add design to tests spec

GiudGiud committed:
Add support for offsetting the 3D cylinder so the center is at the 'position'

GiudGiud committed:
Use the general RZ coordinate system for 2D cylinder components

GiudGiud committed:
Re-gold the two component test due to fixes in rotations.
New one is easier to look at

GiudGiud committed:
Add ability to set a subdomain name for the 3D component cylinder

Commit:
Transition to using the concentric circle mesh generator

GiudGiud reviewed (approved):
@joshuahansel your review should be addressed. I addressed my own review / analysis of the needs for this component (notably adding the named subdomain support)

@joshuahansel do you want to look again at the component parts of this?
@roystgnr do you want to review the mesh rotation parts of this?

GiudGiud committed:
Fix tests requirements/details
No recover for mesh only calcs

GiudGiud commented on framework/include/utils/RotationMatrix.h:
```diff
@@ -83,4 +83,38 @@ rotVec2DToX(const GenericRealVectorValue<is_ad> & vec)
 }
+
+/**
+ * Provides rotatiom matrix for rotating from vec1 to vec2 using Rodrigues' rotation forumula.
```
will address in a later PR

GiudGiud commented:
Verified Rodrigues formula again.
